fix(aio): improve heading-anchor placement

petebacondarwin opened the pull request:
Anchors on the left are problematic where there is an enclosing
box with a different colour background or border, since the
anchor was overlapping the box.

This commit moves it to after the heading where there is much
less chance of it being affected in this way.

The 3rd party plugin that we were using for this anchor injection
did not support providing configuration with the version of remark
that we are using, so this commit reimplements the functionality
more simply in this project.

mary-poppins commented:
The angular.io preview for ad70ba06950170b67f9e069dfd1e6a724ecaf1a3 is available [here][1].

[1]: https://pr16336-ad70ba06950170b67f9e069dfd1e6a724ecaf1a3.ngbuilds.io/

gkalpak commented:
[docs_test][1] fails (it seems that some assertions need to be updated).

[1]: https://travis-ci.org/angular/angular/jobs/225901477

petebacondarwin commented:
Good old tests :-)

petebacondarwin commented:
Fixed the tests.

gkalpak commented:
Not super-excited about having them on the right, but other than that it LGTM (as long as Travis is happy).

mary-poppins commented:
The angular.io preview for 2eb8cf91b5489990e32bbf403c645fa35ab9e8ff is available [here][1].

[1]: https://pr16336-2eb8cf91b5489990e32bbf403c645fa35ab9e8ff.ngbuilds.io/

petebacondarwin commented:
I noticed that @sjtrimble's new PR (#16327) puts a heading anchor on the far right (and doesn't make it invisible when not hovering. This may have been my accident but it looked quite nice.
So I have updated this PR to achieve the same across all headings.

(I also took this as an opportunity to fix the concern that @gkalpak had about the absolute positioning of icons by a generic CSS rule was too broad).

Let me know what you think.

mary-poppins commented:
The angular.io preview for ad27107e356032f2474c204dcf84b974f0cdb0ad is available [here][1].

[1]: https://pr16336-ad27107e356032f2474c204dcf84b974f0cdb0ad.ngbuilds.io/

gkalpak commented:
I am not a huge fan of far-right, always-visible anchors (but I can live with that if we can't find a better solution).

Here's another idea:
If the cases where there is a bordered parent (that messs with the left-side anchor) are limited to a couple known ones, maybe it is worth keeping the left-side anchors (possibly keeping them visible with low opacity when not hovered to make them more discoverable) and special-cases the few situations where there are bordered parents (via CSS styles).

gkalpak reviewed (approved):
Implementation-wise LGTM (with the tests fixed) 👍

petebacondarwin commented:
I just realised this breaks copy button positioning. At least on my phone

On Wed, 26 Apr 2017, 11:29 George Kalpakas, <notifications@github.com>
wrote:

> *@gkalpak* approved this pull request.
>
> Implementation-wise LGTM (with the tests fixed) 👍
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/angular/angular/pull/16336#pullrequestreview-34802309>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAA9JwsKcoAu9nizD3B9m4ArIQpfKy52ks5rzxxygaJpZM4NIehE>
> .
>

gkalpak commented on aio/src/styles/2-modules/_code.scss:
```diff
@@ -88,14 +88,19 @@ aio-code.headed-code {
+
   @media (max-width: 480px) {
     top: 0;
     bottom: 0;
```
This should probably be `right: 0`.

petebacondarwin replied:
I agree.

gkalpak commented:
I don't see any difference with this PR (vs master). Positioning on mobile is probably already broken (see https://github.com/angular/angular/pull/16336/files#r113424809).
(Unless you mean something else.)

petebacondarwin commented:
I fixed the mobile alignment. And I also made it invisible again.

mary-poppins commented:
The angular.io preview for 8dbca6afbf7cfb0770d3b4491b8322bf48e669d2 is available [here][1].

[1]: https://pr16336-8dbca6afbf7cfb0770d3b4491b8322bf48e669d2.ngbuilds.io/

sjtrimble commented:
@petebacondarwin It wasn't my intent to put the anchor on the right :) but I know this might fix some of the header indentation issues for now. The side effect of your updates is having the h1 before element (the line under the h1) extend a bit further out than planned. I will submit an update.

petebacondarwin commented:
Let's check this is what we actually want before you make any more changes

On Wed, 26 Apr 2017, 17:11 Stefanie Fluin, <notifications@github.com> wrote:

> @petebacondarwin <https://github.com/petebacondarwin> It wasn't my intent
> to put the anchor on the right :) but I know this might fix some of the
> header indentation issues for now. The side effect of your updates is
> having the h1 before element (the line under the h1) extend a bit further
> out than planned. I will submit an update.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/angular/angular/pull/16336#issuecomment-297461356>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAA9J-zB5Vx0wyHFi4kj7ahnC4QeMSb7ks5rz2xrgaJpZM4NIehE>
> .
>

petebacondarwin commented:
In our meeting we agreed to move the anchor back to the left side and change the CSS so that if the heading is inside a subsection the anchor is moved further left so that from a user's perspective the anchor always appears in the left gutter of the page.

petebacondarwin committed:
fix(aio): ensure styling of icons is specific to usage

Previously the CSS styling for material icons was too broad and affected
all instances of icons. This commit constrains the position of copy button
icons only to copy buttons.

petebacondarwin committed:
fix(aio): improve positioning of heading-anchors

mary-poppins commented:
The angular.io preview for dc8b9d4423cd66f4bc7b9fa4bfbb4ae79214676e is available [here][1].

[1]: https://pr16336-dc8b9d4423cd66f4bc7b9fa4bfbb4ae79214676e.ngbuilds.io/

wardbell commented:
@petebacondarwin I like the new _anchor-in-the-gutter_.  Looking forward to seeing it work with HTML heading tags as well as MD heading tags. Thanks!

petebacondarwin committed:
build(aio): add new post-process dgeni package

This package will allow us to do complex post-processing
on the HTML that is rendered by Nunjucks.

petebacondarwin committed:
build(aio): move autolink-headings to post-processing

The autolinking is now done on the `renderedContent` which means it also
captures and autolinks headings that were generated outside of markdown.

mary-poppins commented:
The angular.io preview for e5f23aea5f5859a9d4c08c711384a722b487074f is available [here][1].

[1]: https://pr16336-e5f23aea5f5859a9d4c08c711384a722b487074f.ngbuilds.io/

gkalpak commented on aio/tools/transforms/post-process-package/processors/post-process-html.js:
```diff
@@ -0,0 +1,37 @@
+ *
+ * Each plugin is a factory function that will be called with the "rehype" engine as `this`.
+ * The factory should return a function that takes a HAST and returns a `boolean` or `undefined`.
+ * The HAST can be mutated by the this function.
```
_the this_ --> _the_ or _this_

petebacondarwin replied:
The `this` variable

gkalpak replied:
I still think you meant "The HAST can be mutated by this function` (where "this function" is the _"function that takes a HAST and returns a `boolean` or `undefined`"_, which in turn is returned by the plugin factory function) :grin:

petebacondarwin replied:
Oh yeah :-)
